In [231]:
import pandas as pd
from copy import deepcopy
from pymorphy2 import MorphAnalyzer
morph = MorphAnalyzer()

In [232]:
df = pd.read_csv('andic_dicts.csv')

In [233]:
df = df[df.glottocode == 'cham1309']

In [234]:
df.columns

Index(['lang', 'id_word', 'id_meaning', 'id', 'lemma', 'ipa', 'morphology',
       'annotator', 'pos', 'concepticon', 'meaning_ru', 'definition',
       'glottocode', 'reference', 'comments', 'source', 'bor',
       'borrowing_source_language', 'borrowing_source_transcription',
       'borrowing_modified', 'borrowing_annotator', 'language'],
      dtype='object')

In [235]:
df = df[["id_word", "id_meaning", "lemma", "ipa", "morphology", "pos", "meaning_ru"]]

In [236]:
# убираем ударения и номера значений
df.lemma = df.lemma.str.lower().str.replace('́|̀|¹|²|³', '', regex=True)
df.morphology = df.morphology.str.lower().str.replace('́|̀|¹|²|³', '', regex=True)

df.lemma = df.lemma.str.lower().str.replace('á', 'а', regex=True)
df.morphology = df.morphology.str.lower().str.replace('á', 'а', regex=True)
df.lemma = df.lemma.str.lower().str.replace('é', 'е', regex=True)
df.morphology = df.morphology.str.lower().str.replace('é', 'е', regex=True)

# палочка
df.lemma = df.lemma.str.replace('i', '1')
df.morphology = df.morphology.str.replace('i', '1')

# назализация
df.lemma = df.lemma.str.replace('ᴴ', '2')
df.morphology = df.morphology.str.replace('ᴴ', '2')

# дологота
df.lemma = df.lemma.str.replace('(.)̅', r'\1\1', regex=True)
df.morphology = df.morphology.str.replace('(.)̅', r'\1\1', regex=True)

# убираем баги долготы: ссII -> сIсI, ллъъ -> лълъ
df.lemma = df.lemma.str.replace('(.){2}(1|2|ъ){2}', r'\1\2\1\2', regex=True)
df.morphology = df.morphology.str.replace('(.){2}(1|2|ъ){2}', r'\1\2\1\2', regex=True)

In [237]:
df

id_word  id_meaning     lemma               ipa  \
56822        1           1         а                 a   
56823        1           2         а                 a   
56824        2           1        аб               a-b   
56825        3           1   абадал1     ˌa-b-a-d-a-tɬ   
56826        4           1  абадийаб  a-b-a-d-'i-j-a-b   
...        ...         ...       ...               ...   
66826     7081           1   эъусуна     e-ʔ-u-s-u-n-a   
66827     7082           1         э                 e   
66828     7083           1         э                 e   
66829     7084           1     эгьай           e-h-a-j   
66830     7085           1    эт1ула        e-t'-u-l-a   

                                              morphology       pos  \
56822                                      (авл1, аавбе)  noun III   
56823                                      (авл1, аавбе)  noun III   
56824  (ав 1, а 1 1, 1v, ал v, аби мн.ч. кл. чел., аа...      pron   
56825                                                NaN       adv   
56826  (абадййо 1, абадйй 1 1, 1v, абадййел v, -емн.ч...       adj   
...                                                  ...       ...   
66826                                                NaN       NaN   
66827                                         (-л1, -бе)  noun III   
66828                                                NaN       NaN   
66829                                                NaN       NaN   
66830            (эт1в, эт1уда, эт1вбе, эт1ви, эт1удахъ)       NaN   

               meaning_ru  
56822               колос  
56823  початок (кукурузы)  
56824                этот  
56825           ни за что  
56826              вечный  
...                   ...  
66826            краснеть  
66827               кровь  
66828                   а  
66829      межд. ай-ай-ай  
66830     истекать кровью  

[10009 rows x 7 columns]

In [238]:
print(len(df.id_word.drop_duplicates()))
print(len(df.lemma.drop_duplicates()))

7078
6816


In [239]:
print(len(df.drop_duplicates(subset=['id_word'])\
          [~df.pos.isna()]))

2459


C:\Users\USER\AppData\Local\Temp\ipykernel_10276\1168775718.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  print(len(df.drop_duplicates(subset=['id_word'])\


Строк в таблице 10009, при этом только разных слов только 7078, а уникальных лемм - 6819

### POS tagging

In [240]:
def get_pos_tag(word):
    word = str(word)
    return str(morph.parse(word)[0].tag)[:4]

df['my_pos'] = df['meaning_ru'].apply(get_pos_tag)

In [241]:
df.my_pos.unique()

array(['NOUN', 'UNKN', 'ADJF', 'ADJS', 'PRCL', 'ADVB', 'INFN', 'CONJ',
       'PRTF', 'VERB', 'NUMR', 'INTJ', 'GRND', 'PREP', 'PRED', 'NPRO',
       'COMP', 'PNCT', 'NUMB', 'PRTS'], dtype=object)

### Nouns

In [242]:
noun = df[(df.pos.notnull() & df.pos.str.contains("noun")) | (df.pos.isnull() & (df.my_pos == 'NOUN'))]
noun

id_word  id_meaning      lemma                 ipa  \
56822        1           1          а                   a   
56823        1           2          а                   a   
56828        6           1  абазинали  a-b-a-z-'i-n-a-l-i   
56829        7           1    абазине       a-b-a-z-i-n-e   
56830        8           1    абазино      a-b-a-z-'i-n-o   
...        ...         ...        ...                 ...   
66814     7074           2     эъвлъи            e-ʔʷ-ɬ-i   
66817     7076           1    эъугьла         e-ʔ-u-h-l-a   
66819     7077           1   эъвгьула        e-ʔʷ-h-u-l-a   
66827     7082           1          э                   e   
66830     7085           1     эт1ула          e-t'-u-l-a   

                                          morphology       pos  \
56822                                  (авл1, аавбе)  noun III   
56823                                  (авл1, аавбе)  noun III   
56828                                            NaN      noun   
56829                            (-лъил1, абазйнали)   noun II   
56830                              (-суб, абазйнали)    noun I   
...                                              ...       ...   
66814                            (эъвлъйл1, тк. ед.)   noun IV   
66817  (эъугь, эъвгьуда, эъвгьа, эъугьи, эъвгьудахъ)       NaN   
66819                                            NaN       NaN   
66827                                     (-л1, -бе)  noun III   
66830        (эт1в, эт1уда, эт1вбе, эт1ви, эт1удахъ)       NaN   

                   meaning_ru my_pos  
56822                   колос   NOUN  
56823      початок (кукурузы)   UNKN  
56828                 абазины   NOUN  
56829                абазинка   NOUN  
56830                абазинец   NOUN  
...                       ...    ...  
66814                 румянец   NOUN  
66817  красить в красный цвет   NOUN  
66819  красить в красный цвет   NOUN  
66827                   кровь   NOUN  
66830         истекать кровью   NOUN  

[4314 rows x 8 columns]

Нам нужны формы Nom Sg, Nom Pl, основа для падежей типа Dat Sg, типа Gen Sg и всех Pl

Пример

- nom sg: а
- nom pl: а:аавбе
- gen sg: а:ав
- dat sg: а:ав
- obl pl: а:аавбе

In [243]:
noun[noun.morphology.isnull()]

id_word  id_meaning      lemma                 ipa morphology   pos  \
56828        6           1  абазинали  a-b-a-z-'i-n-a-l-i        NaN  noun   
56846       22           1   абхазали    a-b-χ-'a-z-a-l-i        NaN  noun   
56858       32           1     авараг        a-w-'a-r-a-ɡ        NaN  noun   
56890       61           1     адвлъа            a-dʷ-ɬ-a        NaN   NaN   
56897       68           1  адигейали  a-d-i-ɡ-'e-j-a-l-i        NaN  noun   
...        ...         ...        ...                 ...        ...   ...   
66773     7043           1  эч1иазала   e-tʃ'-i-a-z-a-l-a        NaN   NaN   
66779     7046           1  эч1и-сии2      e-tʃ'-'i-s-ĩː        NaN   NaN   
66785     7052           1   эч1-т1уб        e-tʃ'-t'-u-b        NaN   NaN   
66786     7052           1   эч1-т1уб        e-tʃ'-t'-u-b        NaN   NaN   
66819     7077           1   эъвгьула        e-ʔʷ-h-u-l-a        NaN   NaN   

                   meaning_ru my_pos  
56828                 абазины   NOUN  
56846                  абхазы   NOUN  
56858                  пророк   NOUN  
56890                  см. аб   NOUN  
56897                адыгейцы   NOUN  
...                       ...    ...  
66773              две тысячи   NOUN  
66779          в числе первых   NOUN  
66785   в двухкратном размере   NOUN  
66786   в двухкратном размере   NOUN  
66819  красить в красный цвет   NOUN  

[863 rows x 8 columns]

In [244]:
good_noun = noun[noun.morphology.notnull()]
good_noun[~good_noun.morphology.str.contains(",|;")]

id_word  id_meaning               lemma                         ipa  \
57137      276           1            атаквала             a-t-'a-kʷ-a-l-a   
57217      340           1  ахъбак1ул-к1ва2гьа  a-q-b-a-k'-'u-l-k'ʷ-ã-h-a   
57345      445           1                а2та                     ã-t-'a   
57349      448           1                а2чи                     ˌa-tʃ-i   
57409      484           1            баг1араш              b-'a-ʕ-a-r-a-ʃ   
...        ...         ...                 ...                         ...   
66218     6626           1                чупа                    tʃ-u-p-a   
66225     6631           1               чулаб                  tʃ-u-l-a-b   
66392     6752           1              ч1икаб                 tʃ'-i-k-a-b   
66474     6806           1               ч1ухб                   tʃ'-u-χ-b   
66511     6830           1               шалир                   ʃ-a-l-i-r   

      morphology       pos       meaning_ru my_pos  
57137     (гиг.)       NaN       см. ахъилв   NOUN  
57217      (-бе)  noun III             буза   NOUN  
57345     (гиг.)       NaN        см. кIваб   NOUN  
57349     (гиг.)       NaN         см. шахв   NOUN  
57409     (гиг.)       NaN  см. базар-бакье   NOUN  
...          ...       ...              ...    ...  
66218     (гиг.)       NaN         см. хвам   NOUN  
66225     (гиг.)       NaN    см. кIатIураб   NOUN  
66392     (гиг.)       NaN     см. берцинаб   NOUN  
66474     (гиг.)       NaN        см. эхъва   NOUN  
66511     (гиг.)       NaN       см. за̅йли   NOUN  

[110 rows x 8 columns]

In [245]:
good_noun[good_noun.morphology.str.count(',') > 1]

id_word  id_meaning        lemma                ipa  \
56834       12           1    абгунийаб  a-b-ɡ-u-n-i-j-a-b   
57083      227           1       анкъул        a-n-q'-'u-l   
57099      241           1       апараг       a-p-'a-r-a-ɡ   
57133      273           1        ассуб          a-sː-'u-b   
57196      324           1      ахихъа2        a-χ-'i-q-ã   
...        ...         ...          ...                ...   
66732     7011           1        эхвна           e-χʷ-n-a   
66743     7019           1      эхъаана         e-q-aː-n-a   
66768     7038           1  эч1-бахат1в  e-tʃ'-b-a-χ-a-t'ʷ   
66817     7076           1      эъугьла        e-ʔ-u-h-l-a   
66830     7085           1       эт1ула         e-t'-u-l-a   

                                              morphology        pos  \
56834  (айгунийе 1v, алгунййел v, -емн.ч. кл. чел., а...        NaN   
57083                                   (-иб, -л1, -али)  noun I,II   
57099                                 (-иб, -ил1, -забе)  noun I,II   
57133  (асу 1, асв 1 1, 1v, асул v, -емн ч кл. чел., ...        NaN   
57196                         (ахйхъаниб, -л1, ахйхъаме)  noun I,II   
...                                                  ...        ...   
66732             (эхва2, эхуна, эхваме, эхва2, эхунахъ)        NaN   
66743          (эхъан, эхъана, эхъана, эхъаан, эхъанахъ)        NaN   
66768  (и мн. ч. кл. чел., эч1вахат1в 1, эч1йахат1 1 ...        NaN   
66817      (эъугь, эъвгьуда, эъвгьа, эъугьи, эъвгьудахъ)        NaN   
66830            (эт1в, эт1уда, эт1вбе, эт1ви, эт1удахъ)        NaN   

                   meaning_ru my_pos  
56834   см. ащинаб, гьаᴴщинаб   NOUN  
57083                 глухарь   NOUN  
57099                пришелец   NOUN  
57133                  см. аб   NOUN  
57196                 садовод   NOUN  
...                       ...    ...  
66732                    мочь   NOUN  
66743         понуд. от эхъна   NOUN  
66768                 к концу   NOUN  
66817  красить в красный цвет   NOUN  
66830         истекать кровью   NOUN  

[385 rows x 8 columns]

Future work... 
- 863 rows of `noun[noun.morphology.isnull()]` 
- 110 rows of `good_noun[~good_noun.morphology.str.contains(",|;")]`
- 385 rows of `good_noun[good_noun.morphology.str.count(',') > 1]`

In [246]:
good_noun = noun[noun.morphology.notnull()]
good_noun = good_noun[good_noun.morphology.str.contains(",|;")]
good_noun.morphology = good_noun.morphology.str.replace(';', ',', regex=True)  # есть 8 вхождений, где вместо ',' стоит ';'
good_noun.morphology = good_noun.morphology.str.replace('\(|\)', '', regex=True)
good_noun = good_noun[good_noun.morphology.str.count(',') == 1]
good_noun.morphology = good_noun.morphology.str.replace(', ', ',',)
good_noun[['gen_sg', 'nom_pl']] = good_noun.morphology.str.split(",", expand=True)
good_noun

id_word  id_meaning    lemma             ipa        morphology  \
56822        1           1        а               a        авл1,аавбе   
56823        1           2        а               a        авл1,аавбе   
56829        7           1  абазине   a-b-a-z-i-n-e  -лъил1,абазйнали   
56830        8           1  абазино  a-b-a-z-'i-n-o    -суб,абазйнали   
56831        9           1     абас        a-b-'a-s   абасул1,абасвбе   
...        ...         ...      ...             ...               ...   
66808     7069           1      эъа           e-ʔ-a          -ил1,-бе   
66809     7070           1      эъа           e-ʔ-a       эъалал1,-бе   
66813     7074           1   эъвлъи        e-ʔʷ-ɬ-i  эъвлъйл1,тк. ед.   
66814     7074           2   эъвлъи        e-ʔʷ-ɬ-i  эъвлъйл1,тк. ед.   
66827     7082           1        э               e           -л1,-бе   

            pos          meaning_ru my_pos    gen_sg     nom_pl  
56822  noun III               колос   NOUN      авл1      аавбе  
56823  noun III  початок (кукурузы)   UNKN      авл1      аавбе  
56829   noun II            абазинка   NOUN    -лъил1  абазйнали  
56830    noun I            абазинец   NOUN      -суб  абазйнали  
56831  noun III           20 копеек   NOUN   абасул1    абасвбе  
...         ...                 ...    ...       ...        ...  
66808       NaN            сычужина   NOUN      -ил1        -бе  
66809       NaN                туча   NOUN   эъалал1        -бе  
66813   noun IV            краснота   NOUN  эъвлъйл1    тк. ед.  
66814   noun IV             румянец   NOUN  эъвлъйл1    тк. ед.  
66827  noun III               кровь   NOUN       -л1        -бе  

[2956 rows x 10 columns]

In [247]:
# Nom Sg
nom_sg = good_noun[['lemma', 'meaning_ru']]
nom_sg = nom_sg.groupby('lemma', as_index=False).agg({'lemma': 'first', 'meaning_ru': ', '.join})
nom_sg.to_csv('nom_sg.txt', sep='#', index=False, header=False)

Future work...
- 132 rows of `good_noun[~good_noun.gen_sg.str.endswith('л1')]` (-суб, -иб gentivie)

In [248]:
# Gen Sg

# Разводим вариативность по разным строкам
good_noun.gen_sg = good_noun.gen_sg.str.split("/")
good_noun = good_noun.explode("gen_sg").reset_index(drop=True)

# Прописываем основу, где требуется
good_noun.gen_sg = good_noun.gen_sg.str.replace('-|—', '~', regex=True)
good_noun.gen_sg = good_noun.apply(lambda x: x.gen_sg.replace('~', str(x.lemma)), axis=1)

gen_sg = good_noun[['lemma', 'gen_sg']]
gen_sg = gen_sg[gen_sg.gen_sg.str.endswith('л1')]
gen_sg.gen_sg = gen_sg.gen_sg.str.removesuffix('л1')
gen_sg = gen_sg.drop_duplicates()
gen_sg.to_csv('gen_sg.txt', sep=':', index=False, header=False)

In [249]:
# Nom Pl

# Разводим вариативность по разным строкам
good_noun.nom_pl = good_noun.nom_pl.str.split("/")
good_noun = good_noun.explode("nom_pl").reset_index(drop=True)

# Прописываем основу, где требуется
good_noun.nom_pl = good_noun.nom_pl.str.replace('-|—', '~', regex=True)
good_noun.nom_pl = good_noun.apply(lambda x: x.nom_pl.replace('~', str(x.lemma)), axis=1)

# Pl = Sg
good_noun.nom_pl = good_noun.nom_pl.str.replace(r'.+ ед.*', 'тк. ед.', regex=True)
good_noun.nom_pl = good_noun.apply(lambda x: x.nom_pl.replace('тк. ед.', str(x.lemma)), axis=1)

nom_pl = good_noun[['lemma', 'nom_pl']]
nom_pl = nom_pl[~nom_pl.nom_pl.str.contains('собир')]

nom_pl.nom_pl = nom_pl.nom_pl.str.replace(r'дй$', 'ди', regex=True)  # common typo

nom_pl = nom_pl.drop_duplicates()
nom_pl.to_csv('nom_pl.txt', sep=':', index=False, header=False)

Overgeneration
- dat sg: gen_sg + final и -> 0, final у -> в, final у -> 0
- obl pl: nom_pl + final е -> а, final и -> а

In [250]:
# Dat Sg

dat_sg = gen_sg[gen_sg.gen_sg.str.endswith(('и', 'у'))]
dat_sg = deepcopy(dat_sg)
dat_sg.gen_sg = dat_sg.gen_sg.str.removesuffix('и')  # final и -> 0
dat_sg.gen_sg = dat_sg.gen_sg.str.removesuffix('у')  # final у -> 0
print(len(dat_sg))

dat_sg_more = gen_sg[gen_sg.gen_sg.str.endswith('у')]
dat_sg_more = deepcopy(dat_sg_more)
dat_sg_more.gen_sg = dat_sg_more.gen_sg.str.replace('у$', 'в', regex=True)  # final у -> в
print(len(dat_sg_more))

dat_sg = pd.concat([dat_sg, dat_sg_more])
print(len(dat_sg))

dat_sg.to_csv('dat_sg.txt', sep=':', index=False, header=False)

966
439
1405


In [252]:
# Obl Pl

obl_pl = nom_pl[nom_pl.nom_pl.str.endswith(('е', 'а'))]
obl_pl = obl_pl[['lemma', 'nom_pl']]
obl_pl.nom_pl = obl_pl.nom_pl.str.replace(r'[еи]$', 'а', regex=True)  # final е -> а, final и -> а
obl_pl.to_csv('obl_pl.txt', sep=':', index=False, header=False)

### Adjectives

Проблема: тэг ADJ в таблице имеют только 56 прилагательных, хотя их в словаре, очевидно, больше

In [58]:
adj_pos = df[df.pos == "adj"]
adj_pos.shape

(56, 7)

In [ ]:
adj = df[(df.pos.notnull() & df.pos.str.contains("adj")) | (df.pos.isnull() & (df.my_pos == 'ADJF'))]
adj

In [62]:
a_prefix = ('ба', 'ва', 'йа')
e_prefix = ('бе', 'ве', 'йе')
i_prefix = ('би', 'ви', 'йи')

adj['prefix'] = 'no_pref'
adj.loc[adj.lemma.str.startswith(a_prefix), 'prefix'] = 'a-pref'
adj.loc[adj.lemma.str.startswith(e_prefix), 'prefix'] = 'e-pref'
adj.loc[adj.lemma.str.startswith(i_prefix), 'prefix'] = 'i-pref'

for prefix in a_prefix + e_prefix + i_prefix:
    adj.lemma = adj.lemma.str.removeprefix(prefix)

In [63]:
adj['suffix'] = '?'
adj.loc[adj.lemma.str.endswith(('уб', 'ул')), 'suffix'] = 'u-suff'
adj.loc[adj.lemma.str.endswith(('ооб', 'оол')), 'suffix'] = 'oo-suff'
adj.loc[adj.lemma.str.endswith(('аб', 'ел')), 'suffix'] = 'o-suff'

for suffix in ('уб', 'ул', 'аб', 'ел'):
    adj.lemma = adj.lemma.str.removesuffix(suffix)

In [64]:
adj

id_word  id_meaning      lemma                ipa  \
56826        4           1     абадий   a-b-a-d-'i-j-a-b   
56849       25           1     авадан   a-w-a-d-'a-n-a-b   
56860       34           1     авзада        a-w-z-a-d-a   
56869       41           1      агъат       a-ʁ-'a-t-u-b   
56870       41           2      агъат       a-ʁ-'a-t-u-b   
...        ...         ...        ...                ...   
66784     7051           1      эч1л1       e-tʃ'-tɬ-a-b   
66797     7061           1         эш            e-ʃ-u-b   
66810     7071           1  эъалахида  e-ʔ-a-l-a-χ-i-d-a   
66815     7075           1         эъ            e-ʔ-u-b   
66816     7075           2         эъ            e-ʔ-u-b   

                                              morphology  pos   meaning_ru  \
56826  (абадййо 1, абадйй 1 1, 1v, абадййел v, -емн.ч...  adj       вечный   
56849  (авадано 1, авадан 1 1, 1v, аваданел v, -емн. ...  adj      веселый   
56860                                                NaN  adj  общительный   
56869  (агъату 1, агъатв 1 1, 1v, агъатул v, -емн. ч....  adj    неудобный   
56870  (агъату 1, агъатв 1 1, 1v, агъатул v, -емн. ч....  adj    неуклюжий   
...                                                  ...  ...          ...   
66784                                                NaN  NaN       второй   
66797  (эшу 1, эшв 1 1, 1v, эшул v, -емн. ч. кл. чел....  NaN       плохой   
66810                                                NaN  NaN    пасмурный   
66815  (эъу 1, эъв 1 1, 1v, эъул v, -емн. ч. кл. чел....  NaN      красный   
66816                                                NaN  NaN      румяный   

        prefix  suffix  
56826  no_pref  o-suff  
56849  no_pref  o-suff  
56860  no_pref       ?  
56869  no_pref  u-suff  
56870  no_pref  u-suff  
...        ...     ...  
66784  no_pref  o-suff  
66797  no_pref  u-suff  
66810  no_pref       ?  
66815  no_pref  u-suff  
66816  no_pref  u-suff  

[795 rows x 9 columns]

In [65]:
for suffix in ('б', 'л'):
    adj.loc[adj.lemma.str.endswith(('ооб', 'оол')), 'lemma'] = \
    adj.loc[adj.lemma.str.endswith(('ооб', 'оол')), 'lemma'].str.removesuffix(suffix)

Надо ещё что-то сделать с
- 326 словами `adj[adj.suffix == '?']` (непонятные формы прилагательных)

In [66]:
adj_stem = adj[['lemma', 'meaning_ru', 'prefix', 'suffix']]
adj_stem = adj_stem[adj_stem.suffix != '?']
adj_stem

lemma meaning_ru   prefix   suffix
56826     абадий     вечный  no_pref   o-suff
56849     авадан    веселый  no_pref   o-suff
56869      агъат  неудобный  no_pref   u-suff
56870      агъат  неуклюжий  no_pref   u-suff
56898     адийоо  маленький  no_pref  oo-suff
...          ...        ...      ...      ...
66764  эч1ац1ал1  двадцатый  no_pref   o-suff
66784      эч1л1     второй  no_pref   o-suff
66797         эш     плохой  no_pref   u-suff
66815         эъ    красный  no_pref   u-suff
66816         эъ    румяный  no_pref   u-suff

[466 rows x 4 columns]

In [67]:
adj_stem = adj_stem.groupby(['lemma', 'prefix', 'suffix'], as_index=False).agg({'meaning_ru': ', '.join})

In [68]:
adj_stem = adj_stem[adj_stem.lemma != '']

In [69]:
adj_stem

lemma   prefix  suffix           meaning_ru
1    а2зигьац1  no_pref  o-suff          белоснежный
2         а2сс  no_pref  o-suff               теплый
3          а2т  no_pref  o-suff  молодой, некорепший
4       абадий  no_pref  o-suff               вечный
5       авадан  no_pref  o-suff              веселый
..         ...      ...     ...                  ...
306         эх  no_pref  u-suff                полый
307  эч1ац1ал1  no_pref  o-suff            двадцатый
308      эч1л1  no_pref  o-suff               второй
309         эш  no_pref  u-suff               плохой
310         эъ  no_pref  u-suff     красный, румяный

[310 rows x 4 columns]

In [70]:
adj_stem['result'] = ''
adj_stem.result = adj_stem.lemma + '[' + adj_stem.prefix + ',' + adj_stem.suffix + ']' + '  # ' + adj_stem.meaning_ru

In [71]:
adj_stem.to_csv('adj_stem.txt', columns=['result'], sep=' ', index=False, header=False)

### Verbs

Опять проблема с отсутствием тегов

In [9]:
df[df.meaning_ru == 'стрелять']

id_word  id_meaning        lemma                   ipa  \
57793      730           2      бахаала          b-a-χ-'a-l-a   
57816      742           2    бахихаала      b-a-χ-i-χ-'a-l-a   
62819     4051           1     кьадаала        tɬ'-a-d-aː-l-a   
62824     4054           1  кьакьадаала  tɬ'-a-tɬ'-a-d-aː-l-a   
66443     6784           2       ч1иина            tʃ'-'i-n-a   

                                              morphology  pos meaning_ru  
57793  (к мн. ч. кл. чел., ва-1, й-1-1, 1v, v и мн. ч...  NaN   стрелять  
57816  (мн. ч. кл. чел., ва-1, й-1-1, 1v, v и мн. ч к...  NaN   стрелять  
62819    (кьадал, кьадала, кьадала, кьадаалв, кьадалахъ)  NaN   стрелять  
62824  (кьакьадал, кьакьадала, кьакьадала, кьакьадаал...  NaN   стрелять  
66443                                                NaN  NaN   стрелять

In [10]:
verb_pos = df[df.pos == 'verb']
verb_pos

id_word  id_meaning         lemma                  ipa  \
56850       26           1    авада2лъла     a-w-a-d-'a-ɬ-l-a   
56851       27           1  авада2лъиила  a-w-a-d-'a-ɬ-iː-l-a   
56861       35           1       авзаала         a-w-z-aː-l-a   
56863       37           1         авзла            a-w-z-l-a   
56864       37           2         авзла            a-w-z-l-a   
...        ...         ...           ...                  ...   
57673      669           2   басралъиила  b-a-s-r-'a-ɬ-iː-l-a   
57677      671           1   бассвассвна   b-ˌa-sːʷ-a-sːʷ-n-a   
57678      672           1  бассвассууна  b-a-sːʷ-a-sː-'u-n-a   
57680      674           1       бассвна         b-ˌa-sːʷ-n-a   
57681      675           1      бассууна        b-a-sː-'u-n-a   

                                              morphology   pos     meaning_ru  
56850  (авада2лъ, авада2лъда, авада2лъбе, авада2лъи, ...  verb  стать веселым  
56851                                                NaN  verb  стать веселым  
56861          (авзал, авзала, авзала, авзаал, авзалахъ)  verb      смешивать  
56863               (авза, авзда, авзабе, авза, авздахъ)  verb    смешиваться  
56864               (авза, авзда, авзабе, авза, авздахъ)  verb       общаться  
...                                                  ...   ...            ...  
57673                                                NaN  verb    изувечиться  
57677  (басуса2, басвасуна, басуса2, басвасви2, басва...  verb     складывать  
57678                                                NaN  verb     складывать  
57680          (басу2, басуна, басва2, басви2, басунахъ)  verb     складывать  
57681                                                NaN  verb     складывать  

[280 rows x 7 columns]

In [ ]:
verb = df[(df.pos.notnull() & df.pos.str.contains("verb")) | (df.pos.isnull() & (df.my_pos == 'INFN'))]
verb

In [16]:
verb[verb.morphology.isna()]

id_word  id_meaning         lemma                  ipa morphology  \
56851       27           1  авада2лъиила  a-w-a-d-'a-ɬ-iː-l-a        NaN   
56865       38           1       авзуула         a-w-z-uː-l-a        NaN   
56866       38           2       авзуула         a-w-z-uː-l-a        NaN   
56959      127           1     аквакууна      a-kʷ-a-k-uː-n-a        NaN   
56964      132           1        акуула           a-k-'u-l-a        NaN   
...        ...         ...           ...                  ...        ...   
66820     7077           2      эъвгьула         e-ʔʷ-h-u-l-a        NaN   
66822     7078           2       эъулъла          e-ʔ-u-ɬ-l-a        NaN   
66823     7079           1      эъвлъула         e-ʔʷ-ɬ-u-l-a        NaN   
66824     7079           2      эъвлъула         e-ʔʷ-ɬ-u-l-a        NaN   
66826     7081           1       эъусуна        e-ʔ-u-s-u-n-a        NaN   

        pos     meaning_ru  
56851  verb  стать веселым  
56865  verb    смешиваться  
56866  verb       общаться  
56959  verb       садиться  
56964  verb           жать  
...     ...            ...  
66820   NaN       румянить  
66822   NaN  разрумяниться  
66823   NaN       краснеть  
66824   NaN  разрумяниться  
66826   NaN       краснеть  

[1231 rows x 7 columns]

In [13]:
verb.morphology[57355]

'(и мн. ч. кл. чел , вагъа-вачардаала 1, йагъа-йачардаал 1 1, 1v, v и мн. ч. кл. вещи; багъа-бачардал, багъа- бачардала, багъа-бачардала, багъа-бачардаалв, багъа-бачардалахъ)'

In [14]:
# распарсить часть про классные показатели до ";" нереально
verb.morphology = verb.morphology.str.replace('.*;\s?', '', regex=True)  

In [15]:
verb.morphology[57355]

'багъа-бачардал, багъа- бачардала, багъа-бачардала, багъа-бачардаалв, багъа-бачардалахъ)'

In [98]:
verb.morphology = verb.morphology.str.replace('.', ',', regex=False)

Надо ещё что-то сделать с
- 1348 словами `verb[verb.morphology.str.count(',') != 4]` (либо не глаголы, либо не указаны (все) формы)

In [99]:
verb[verb.morphology.str.count(',') != 4]

id_word  id_meaning         lemma                  ipa morphology  \
56851       27           1  авада2лъиила  a-w-a-d-'a-ɬ-iː-l-a        NaN   
56865       38           1       авзуула         a-w-z-uː-l-a        NaN   
56866       38           2       авзуула         a-w-z-uː-l-a        NaN   
56959      127           1     аквакууна      a-kʷ-a-k-uː-n-a        NaN   
56964      132           1        акуула           a-k-'u-l-a        NaN   
...        ...         ...           ...                  ...        ...   
66820     7077           2      эъвгьула         e-ʔʷ-h-u-l-a        NaN   
66822     7078           2       эъулъла          e-ʔ-u-ɬ-l-a        NaN   
66823     7079           1      эъвлъула         e-ʔʷ-ɬ-u-l-a        NaN   
66824     7079           2      эъвлъула         e-ʔʷ-ɬ-u-l-a        NaN   
66826     7081           1       эъусуна        e-ʔ-u-s-u-n-a        NaN   

        pos     meaning_ru  
56851  verb  стать веселым  
56865  verb    смешиваться  
56866  verb       общаться  
56959  verb       садиться  
56964  verb           жать  
...     ...            ...  
66820   NaN       румянить  
66822   NaN  разрумяниться  
66823   NaN       краснеть  
66824   NaN  разрумяниться  
66826   NaN       краснеть  

[1348 rows x 7 columns]

In [100]:
verb.morphology = verb.morphology.str.replace('\(|\)', '', regex=True)
verb.morphology = verb.morphology.str.replace(', ', ',')
verb.morphology = verb.morphology.str.replace('- ', '-')
good_verb = verb[verb.morphology.str.count(',') == 4]
good_verb = deepcopy(good_verb)
good_verb[['pst', 'fut', 'imp', 'conv_pst', 'conv_prs']] = good_verb.morphology.str.split(",", expand=True)

In [101]:
good_verb = deepcopy(good_verb)
for form in ['lemma', 'pst', 'fut', 'imp', 'conv_pst', 'conv_prs']:
    good_verb[form] = good_verb[form].str.split("/")
    good_verb = good_verb.explode(form).reset_index(drop=True)

* ба - a-pref, n-pref
* бе - e-pref
* би - ia-pref, i-pref
* бу - u-pref
* __ - no_pref

In [102]:
good_verb[good_verb.lemma.str.startswith(('ба', 'бе', 'би', 'бу'))]

id_word  id_meaning             lemma                           ipa  \
58        454           1  багъа-бачардаала  b-a-ʁ-a-b-a-tʃ-'a-r-d-aː-l-a   
59        454           2  багъа-бачардаала  b-a-ʁ-a-b-a-tʃ-'a-r-d-aː-l-a   
60        456           1   багъа-бачариила    b-a-ʁ-a-b-a-tʃ-'a-r-iː-l-a   
61        456           2   багъа-бачариила    b-a-ʁ-a-b-a-tʃ-'a-r-iː-l-a   
62        457           1         багъагъна                b-ˌa-ʁ-a-ʁ-n-a   
...       ...         ...               ...                           ...   
1150     1739           2           буъаана                  b-u-ʔ-'a-n-a   
1151     1741           1             буъла                    b-'u-ʔ-l-a   
1152     1743           1             буъна                    b-'u-ʔ-n-a   
1153     1743           2             буъна                    b-'u-ʔ-n-a   
1154     1743           3             буъна                    b-'u-ʔ-n-a   

                                             morphology   pos   meaning_ru  \
58    багъа-бачардал,багъа-бачардала,багъа-бачардала...  verb     сдвигать   
59    багъа-бачардал,багъа-бачардала,багъа-бачардала...  verb  расшевелить   
60    багъа-бачарид,багъа-бачарда,багъа-бачаридбе,ба...  verb    двигаться   
61    багъа-бачарид,багъа-бачарда,багъа-бачаридбе,ба...  verb    двинуться   
62    багъигъа2,багъагъйна,багъагъме,багъйгъа2,багъа...  verb     задевать   
...                                                 ...   ...          ...   
1150               буъан,буъана,буъана,буъаанв,буъанахъ   NaN     доводить   
1151                     буъа,буъда,буъабе,буъа,буъдахъ   NaN     сползать   
1152                 буъа2,буъуна,буъаме,буъа2,буъунахъ   NaN    вырастать   
1153                 буъа2,буъуна,буъаме,буъа2,буъунахъ   NaN      хватать   
1154                 буъа2,буъуна,буъаме,буъа2,буъунахъ   NaN        расти   

                 pst              fut              imp          conv_pst  \
58    багъа-бачардал  багъа-бачардала  багъа-бачардала  багъа-бачардаалв   
59    багъа-бачардал  багъа-бачардала  багъа-бачардала  багъа-бачардаалв   
60     багъа-бачарид    багъа-бачарда  багъа-бачаридбе    багъа-бачарйдв   
61     багъа-бачарид    багъа-бачарда  багъа-бачаридбе    багъа-бачарйдв   
62         багъигъа2       багъагъйна        багъагъме         багъйгъа2   
...              ...              ...              ...               ...   
1150           буъан           буъана           буъана           буъаанв   
1151            буъа            буъда           буъабе              буъа   
1152           буъа2           буъуна           буъаме             буъа2   
1153           буъа2           буъуна           буъаме             буъа2   
1154           буъа2           буъуна           буъаме             буъа2   

               conv_prs  
58    багъа-бачардалахъ  
59    багъа-бачардалахъ  
60     багъа-бача рдахъ  
61     багъа-бача рдахъ  
62         багъагъйнахъ  
...                 ...  
1150           буъанахъ  
1151            буъдахъ  
1152           буъунахъ  
1153           буъунахъ  
1154           буъунахъ  

[1084 rows x 12 columns]

In [103]:
good_verb['prefix'] = '[no_pref]'
good_verb.loc[good_verb.lemma.str.startswith('ба'), 'prefix'] = '[a-pref,n-pref]'
good_verb.loc[good_verb.lemma.str.startswith('бе'), 'prefix'] = '[e-pref]'
good_verb.loc[good_verb.lemma.str.startswith('би'), 'prefix'] = '[ia-pref,i-pref]'
good_verb.loc[good_verb.lemma.str.startswith('бу'), 'prefix'] = '[u-pref]'

for prefix in ('ба', 'бе', 'би', 'бу'):
    for form in ('lemma', 'pst', 'fut', 'imp', 'conv_pst', 'conv_prs'):
        good_verb[form] = good_verb[form].str.removeprefix(prefix)

In [104]:
# отсеивается 1 слово
good_verb = good_verb[~good_verb.morphology.str.contains('мн ч')]

In [105]:
good_verb = good_verb.groupby(['lemma', 'prefix', 'pst', 'fut', 'imp', 'conv_pst', 'conv_prs'], as_index=False).\
    agg({'meaning_ru': ', '.join})

In [106]:
good_verb

lemma     prefix      pst        fut        imp  conv_pst  \
0      а2ссаала  [no_pref]    а2сал     а2сала     а2сала   а2саалв   
1     а2ссагьла  [no_pref]   а2сагь  а2сагьида    а2сагьа   а2сагьи   
2     а2ссаквла  [no_pref]   а2сакв   а2сакуда   а2саквбе   а2сакви   
3     а2ссалъла  [no_pref]   а2салъ  а2салъида   а2салъбе   а2салъи   
4     а2ссассла  [no_pref]    а2сас   а2сасида     а2сиса    а2саси   
...         ...        ...      ...        ...        ...       ...   
1419  эч1бакъла  [no_pref]  эч1бакъ  эч1бакъда  эч1бакъбе  эч1бакъи   
1420    эшугьла  [no_pref]    эшугь    эшугьда     эшвгьа    эшугьи   
1421    эшулъла  [no_pref]    эшулъ    эшлъуда    эшулъбе    эшулъи   
1422    эъулъла  [no_pref]    эъул1   эъвлъуда    эъулъбе    эъулъи   
1423     эъусна  [no_pref]   эъуса2    эъусуна    эъусаме    эъуса2   

         conv_prs        meaning_ru  
0        а2салахъ          надевать  
1     а2сагьидахъ          утеплять  
2      а2сакудахъ           греться  
3     а2салъидахъ           теплеть  
4      а2сасидахъ  надевать на себя  
...           ...               ...  
1419  эч1бакъдахъ       удваиваться  
1420   эшвгьудахъ          ухудшать  
1421    эшлъудахъ        ухудшаться  
1422   эъвлъудахъ          краснеть  
1423    эъусунахъ          краснеть  

[1424 rows x 8 columns]

In [110]:
good_verb = good_verb.drop_duplicates()

for form in ('lemma', 'pst', 'fut', 'imp', 'conv_pst', 'conv_prs'):
    good_verb.to_csv(form + '.txt', columns=[form, 'prefix', 'meaning_ru'], sep='#', index=False, header=False)

### Adverbs

In [42]:
df.my_pos.unique()

array(['NOUN', 'UNKN', 'ADJF', 'ADJS', 'PRCL', 'ADVB', 'INFN', 'CONJ',
       'PRTF', 'VERB', 'NUMR', 'INTJ', 'GRND', 'PREP', 'PRED', 'NPRO',
       'COMP', 'PNCT', 'NUMB', 'PRTS'], dtype=object)

In [106]:
adv = df[(df.pos == 'adv') | (df.pos.isnull() & (df.my_pos == 'ADVB'))]
adv

id_word  id_meaning        lemma                ipa morphology  pos  \
56825        3           1      абадал1      ˌa-b-a-d-a-tɬ        NaN  adv   
56862       36           1  авза-гъавза  a-w-z-a-ʁ-a-w-z-a        NaN  adv   
56877       48           1        адакь         a-d-'a-tɬ'        NaN  adv   
56879       50           1     адакьоол     a-d-a-tɬ'-'o-l        NaN  adv   
56880       51           1         адал            a-d-a-l        NaN  adv   
...        ...         ...          ...                ...        ...  ...   
66791     7057           1      эшвилъа        e-ʃʷ-'i-ɬ-a        NaN  NaN   
66802     7066           1         эщда           e-ʃː-d-a        NaN  NaN   
66803     7066           2         эщда           e-ʃː-d-a        NaN  NaN   
66804     7067           1       эщдада       e-ʃː-d-a-d-a        NaN  NaN   
66805     7067           2       эщдада       e-ʃː-d-a-d-a        NaN  NaN   

       meaning_ru my_pos  
56825   ни за что   ADJS  
56862  вперемешку   ADVB  
56877         так   CONJ  
56879   настолько   ADVB  
56880        сюда   ADVB  
...           ...    ...  
66791       назло   ADVB  
66802       очень   ADVB  
66803     слишком   ADVB  
66804       очень   ADVB  
66805     слишком   ADVB  

[548 rows x 8 columns]

In [80]:
adv[adv.lemma.str.startswith('бе')]

id_word  id_meaning          lemma                     ipa  \
58107      890           1      бегъусса2          b-'e-ʁ-u-sː-ã   
58113      894           1    бегьа2ц1уда     b-'e-h-ã-ts'-u-d-a   
58200      948           2      бекьал1а2        b-e-tɬ'-'a-tɬ-ã   
58207      955           1     бекьикьида    b-e-tɬ'-i-tɬ'-'i-d-a   
58293     1005           1       берци2го        b-'e-r-ts-ĩ-ɡ-o   
58294     1005           2       берци2го        b-'e-r-ts-ĩ-ɡ-o   
58295     1005           3       берци2го        b-'e-r-ts-ĩ-ɡ-o   
58451     1059           1          бехха               b-'e-χː-a   
58452     1059           2          бехха               b-'e-χː-a   
58463     1067           1       беххдале         b-e-χː-d-a-l-'e   
58475     1074           2    беххилалъла     b-e-χː-'i-l-a-ɬ-l-a   
58477     1075           2  беххилалъиила  b-e-χː-'i-l-a-ɬ-iː-l-a   
58478     1076           1         беххоо              b-'e-χː-oː   
58480     1077           1       беххооли          b-'e-χː-oː-l-i   
58526     1091           1      бех1еегъв             b-e-ħ-'e-ʁʷ   
58570     1122           1       беч1удда         b-'e-tʃ'-u-dː-a   
58571     1122           2       беч1удда         b-'e-tʃ'-u-dː-a   
58591     1138           1         беегъв                 b-eː-ʁʷ   
58592     1138           2         беегъв                 b-eː-ʁʷ   
58593     1138           3         беегъв                 b-eː-ʁʷ   
58594     1139           1   беегъв-бегъв          b-'e-ʁʷ-b-e-ʁʷ   
58595     1140           1       беегъвда             b-'e-ʁʷ-d-a   
58596     1141           1    беегъвлълъи            b-'e-ʁʷ-ɬː-i   
58597     1141           2    беегъвлълъи            b-'e-ʁʷ-ɬː-i   
58600     1144           1          беехъ                  b-eː-q   
58602     1145           1        беехъал              b-'e-q-a-l   
58604     1147           1         беехъе                b-'e-q-e   
58605     1147           2         беехъе                b-'e-q-e   
58611     1153           1         беехъо                b-'e-q-o   
58614     1155           1    беехъо-бехъ         b-eː-q-'o-b-e-q   

                                              morphology  pos   meaning_ru  \
58107  ( мн.ч. кл чел , ве-1, -й-1-1, 1v, v и мн. ч. ...  NaN     зачастую   
58113                                                NaN  NaN      сто раз   
58200                                                NaN  NaN        рядом   
58207                                                NaN  NaN    по восемь   
58293                                                NaN  NaN  по-хорошему   
58294                                                NaN  NaN     достойно   
58295                                                NaN  NaN    аккуратно   
58451  (и мн. ч. кл. чел., ве-1, й-1-1, 1v, v и мн. ч...  NaN        назад   
58452  (и мн. ч. кл. чел., ве-1, й-1-1, 1v, v и мн. ч...  NaN        прочь   
58463  (и мн. ч кл чел , ве- 1, й-1-1, 1v, v и мн. ч....  NaN        внизу   
58475  (и мн ч. кл. чел., ве-1, й-1-1, 1v, v ими. ч. ...  NaN   стать выше   
58477                                                NaN  NaN   стать выше   
58478  (и мн. ч кл чел, ве-1, й-1-1, 1v, v и мн. ч кл...  NaN       поздно   
58480                                                NaN  NaN       поздно   
58526     (и мн ч. кл. чел., йе- 1v, v и мн ч. кл. вещи)  NaN        много   
58570  (и мн. ч кл. чел., ве- 1, й-1-1, 1v, v и мн ч....  NaN    бесплатно   
58571  (и мн. ч кл. чел., ве- 1, й-1-1, 1v, v и мн ч....  NaN        пусто   
58591  (и мн. ч кл. чел., ве-1, й-1-1, 1v, v и мн. ч....  NaN        много   
58592  (и мн. ч кл. чел., ве-1, й-1-1, 1v, v и мн. ч....  NaN      слишком   
58593  (и мн. ч кл. чел., ве-1, й-1-1, 1v, v и мн. ч....  NaN      гораздо   
58594  (и мн. ч кл чел., йегъв- йегъв 1v, v и мн. ч. ...  NaN      помногу   
58595  (и мн. ч кл. чел, ве- 1, й-1-1, 1v, v и мн ч к...  NaN      слишком   
58596  (и мн. ч кл чел., ве-1, й-1-1, 1v, v и мн ч кл...  NaN        долг

In [107]:
# there are no adverbs starting with ва, ве, ви / йа, йе, йи

adv = deepcopy(adv)

adv['prefix'] = 'no_pref'
adv.loc[adv.lemma.str.startswith('ба') & adv.morphology.notnull(), 'prefix'] = 'a-pref'
adv.loc[adv.lemma.str.startswith('бе') & adv.morphology.notnull(), 'prefix'] = 'e-pref'
adv.loc[adv.lemma.str.startswith('би') & adv.morphology.notnull(), 'prefix'] = 'i-pref'

for prefix in ('ба', 'бе', 'би'):
    adv.lemma = adv.lemma.str.removeprefix(prefix)

In [93]:
adv[adv.prefix != 'no_pref']

id_word  id_meaning        lemma                   ipa  \
57497      537           1      кьал1а2      b-a-tɬ'-'a-tɬ-ã   
57502      540           1        кьахъ          b-a-tɬ'-'a-q   
57507      544           1   кьа2-ссии2    b-'a-tɬ'-ã-sː-ĩː   
57510      547           1   кьбат1влоо  b-a-tɬ'-b-a-t'ʷ-l-'o   
57684      678           1     с1с1агьи        b-'a-s'ː-a-h-i   
57685      678           2     с1с1агьи        b-'a-s'ː-a-h-i   
57690      681           1       с1с1ад          b-'a-s'ː-a-d   
57771      721           1  хат1в-ссии2     b-a-χ-a-t'ʷ-sː-'i   
58040      848           1        щагьи         b-'a-ʃː-a-h-i   
58041      849           1       щагьла       b-'a-ʃː-a-h-l-a   
58045      851           1          щад           b-'a-ʃː-a-d   
58107      890           1      гъусса2        b-'e-ʁ-u-sː-ã   
58451     1059           1          хха             b-'e-χː-a   
58452     1059           2          хха             b-'e-χː-a   
58463     1067           1       ххдале       b-e-χː-d-a-l-'e   
58475     1074           2    ххилалъла   b-e-χː-'i-l-a-ɬ-l-a   
58478     1076           1         ххоо            b-'e-χː-oː   
58526     1091           1      х1еегъв           b-e-ħ-'e-ʁʷ   
58570     1122           1       ч1удда       b-'e-tʃ'-u-dː-a   
58571     1122           2       ч1удда       b-'e-tʃ'-u-dː-a   
58591     1138           1         егъв               b-eː-ʁʷ   
58592     1138           2         егъв               b-eː-ʁʷ   
58593     1138           3         егъв               b-eː-ʁʷ   
58594     1139           1   егъв-бегъв        b-'e-ʁʷ-b-e-ʁʷ   
58595     1140           1       егъвда           b-'e-ʁʷ-d-a   
58596     1141           1    егъвлълъи          b-'e-ʁʷ-ɬː-i   
58597     1141           2    егъвлълъи          b-'e-ʁʷ-ɬː-i   
58602     1145           1        ехъал            b-'e-q-a-l   
58604     1147           1         ехъе              b-'e-q-e   
58605     1147           2         ехъе              b-'e-q-e   
58611     1153           1         ехъо              b-'e-q-o   
58830     1301           5          сна            b-ˌi-s-n-a   
59181     1455           2          ило             b-iː-l-'o   

                                              morphology  pos      meaning_ru  \
57497  (и мн. ч. кл. чел., ва-1, й-1-1, 1v, v и мн. ч...  adv         в талии   
57502  (и мн. ч. кл. чел., ва-1, й-1-1, 1v, v и мн. ч...  adv      в пояснице   
57507  (и мн. ч. кл. чел., ва-1, й-1-1, 1v, v и мн. ч...  adv         изредка   
57510                     (вакьвит1влоо 1, йакьйит1влоо)  adv       до отвала   
57684  (и мн ч кл чел , ва-1, й-1-1, 1v, v и мн. ч кл...  adv         дочиста   
57685  (и мн ч кл чел , ва-1, й-1-1, 1v, v и мн. ч кл...  adv       подчистую   
57690          (мн. ч кл чел., йа-1v, v и мн. ч кл вещи)  NaN          сплошь   
57771  (и мн ч. кл. чел., ва-1, й-1-1, 1v, v и мн. ч....  NaN          подряд   
58040    (и мн. ч. кл. чел., йа-1v, v и мн. ч. кл. вещи)  NaN         поровну   
58041  ( мн.ч кл чел., ва-1, й-1-1, 1v, v и мн ч. кл....  NaN  делить поровну   
58045  (ими. ч. кл. чел., ва-1, й-1-1, 1v, v и мн. ч....  NaN       одинаково   
58107  ( мн.ч. кл чел , ве-1, -й-1-1, 1v, v и мн. ч. ...  NaN        зачастую   
58451  (и мн. ч. кл. чел., ве-1, й-1-1, 1v, v и мн. ч...  NaN           назад   
58452  (и мн. ч. кл. чел., ве-1, й-1-1, 1v, v и мн. ч...  NaN           прочь   
58463  (и мн. ч кл чел , ве- 1, й-1-1, 1v, v и мн. ч....  NaN           внизу   
58475  (и мн ч. кл. чел., ве-1, й-1-1, 1v, v ими. ч. ...  NaN      стать выше   
58478  (и мн. ч кл чел, ве-1, й-1-1, 1v, v и мн. ч кл...  NaN          поздно   
58526     (и мн ч. кл. чел., йе- 1v, v и мн ч. кл. вещи)  NaN           много   
58570  (и мн. ч кл. чел., ве- 1, й-1-1, 1v, v и мн ч....  NaN       бесплатно   
58571  (и мн. ч кл. чел., ве- 1, й-1-1, 1v, v и мн ч....  NaN           пусто   
58591  (и мн. ч кл. чел., ве-1, й-1-1, 1v, v и мн. ч....  NaN           много   
58592  (и мн. 

In [109]:
adv = adv.groupby(['lemma', 'prefix'], as_index=False).agg({'meaning_ru': ', '.join})

In [110]:
adv['result'] = ''
adv.result = adv.lemma + '[' + adv.prefix + ']' + '  # ' + adv.meaning_ru

In [111]:
adv

lemma   prefix                meaning_ru  \
0    а2л1ц1уда  no_pref                    шестью   
1         а2сс  no_pref             сначала, пока   
2       а2ссаб  no_pref                     тепло   
3     а2ссабда  no_pref  в теплом виде, пока цело   
4    а2ссл1ама  no_pref                     порой   
..         ...      ...                       ...   
478  эч1ич1ида  no_pref                    по два   
479  эч1ла-эч1  no_pref              больше всего   
480    эшвилъа  no_pref                     назло   
481       эщда  no_pref            очень, слишком   
482     эщдада  no_pref            очень, слишком   

                                            result  
0                     а2л1ц1уда[no_pref]  # шестью  
1                   а2сс[no_pref]  # сначала, пока  
2                         а2ссаб[no_pref]  # тепло  
3    а2ссабда[no_pref]  # в теплом виде, пока цело  
4                      а2ссл1ама[no_pref]  # порой  
..                                             ...  
478                   эч1ич1ида[no_pref]  # по два  
479             эч1ла-эч1[no_pref]  # больше всего  
480                      эшвилъа[no_pref]  # назло  
481                эщда[no_pref]  # очень, слишком  
482              эщдада[no_pref]  # очень, слишком  

[483 rows x 4 columns]

In [112]:
adv.to_csv('adv.txt', columns=['result'], sep=' ', index=False, header=False)

### Other

In [8]:
df.pos.unique()

array(['noun III', 'pron', 'adv', 'adj', nan, 'noun', 'noun II', 'noun I',
       'particle', 'noun IV', 'noun V', 'verb', 'num', 'expression',
       'adv?', 'intj', 'conj', 'noun I,II', 'noun VI', 'excl?', 'post',
       'num?', 'onomatope', 'morpheme', 'noun I,II,III', 'noun I,III',
       'noun II,IV'], dtype=object)

In [13]:
print(', '.join(sorted(map(str, df.pos.unique()))))

adj, adv, adv?, conj, excl?, expression, intj, morpheme, nan, noun, noun I, noun I,II, noun I,II,III, noun I,III, noun II, noun II,IV, noun III, noun IV, noun V, noun VI, num, num?, onomatope, particle, post, pron, verb


In [18]:
df[df.pos == 'morpheme']

id_word  id_meaning lemma      ipa morphology       pos meaning_ru
57494      535           1  бакь  b-a-tɬ'        NaN  morpheme   середина

In [128]:
df.my_pos.unique()

array(['NOUN', 'UNKN', 'ADJF', 'ADJS', 'PRCL', 'ADVB', 'INFN', 'CONJ',
       'PRTF', 'VERB', 'NUMR', 'INTJ', 'GRND', 'PREP', 'PRED', 'NPRO',
       'COMP', 'PNCT', 'NUMB', 'PRTS'], dtype=object)

In [162]:
other_pos = ['particle', 'conj', 'intj', 'onomatope', 'post', 'excl?', 'expression']
other_my_pos = ['PREP', 'CONJ', 'INTJ', 'PRCL', 'PRED']

others = df[df.pos.isin(other_pos) | (df.pos.isnull() & (df.my_pos.isin(other_my_pos)))]

In [157]:
others

id_word  id_meaning      lemma              ipa  \
56832       10           1      абаай         a-b-'a-j   
56835       13           1        абе           a-b-'e   
56836       14           1  абек1оохъ   a-b-'e-k'-oː-q   
56837       15           1    абийаай    a-b-'i-j-aː-j   
56841       19           1       абоо           a-b-'o   
...        ...         ...        ...              ...   
66651     6952           1  экъвадала  e-q'ʷ-a-d-a-l-a   
66671     6964           1       эла2           e-l-ã   
66806     7068           1   эщлъихъа    e-ʃː-ɬ-'i-q-a   
66807     7068           2   эщлъихъа    e-ʃː-ɬ-'i-q-a   
66828     7083           1          э                e   

                                              morphology       pos meaning_ru  \
56832                                                NaN  particle        вот   
56835  (аве 1, ай 1 1, 1v, алй v, абй мн.ч. кл.чел., ...  particle        вот   
56836  (авек1оохъ 1, айек1оох 1 1, 1v, алйк1оохъ v,аб...  particle     вот же   
56837                                                NaN  particle        вот   
56841                                                NaN  particle        вот   
...                                                  ...       ...        ...   
66651                                                NaN       NaN        нет   
66671                                                NaN       NaN          в   
66806                                                NaN       NaN      плохо   
66807                                                NaN       NaN      плохо   
66828                                                NaN       NaN          а   

      my_pos  
56832   PRCL  
56835   PRCL  
56836   NOUN  
56837   PRCL  
56841   PRCL  
...      ...  
66651   PRED  
66671   PREP  
66806   PRED  
66807   PRED  
66828   CONJ  

[288 rows x 8 columns]

In [163]:
others = deepcopy(others)

others.pos = others['pos'].fillna(others['my_pos'])

In [164]:
others

id_word  id_meaning      lemma              ipa  \
56832       10           1      абаай         a-b-'a-j   
56835       13           1        абе           a-b-'e   
56836       14           1  абек1оохъ   a-b-'e-k'-oː-q   
56837       15           1    абийаай    a-b-'i-j-aː-j   
56841       19           1       абоо           a-b-'o   
...        ...         ...        ...              ...   
66651     6952           1  экъвадала  e-q'ʷ-a-d-a-l-a   
66671     6964           1       эла2           e-l-ã   
66806     7068           1   эщлъихъа    e-ʃː-ɬ-'i-q-a   
66807     7068           2   эщлъихъа    e-ʃː-ɬ-'i-q-a   
66828     7083           1          э                e   

                                              morphology       pos meaning_ru  \
56832                                                NaN  particle        вот   
56835  (аве 1, ай 1 1, 1v, алй v, абй мн.ч. кл.чел., ...  particle        вот   
56836  (авек1оохъ 1, айек1оох 1 1, 1v, алйк1оохъ v,аб...  particle     вот же   
56837                                                NaN  particle        вот   
56841                                                NaN  particle        вот   
...                                                  ...       ...        ...   
66651                                                NaN      PRED        нет   
66671                                                NaN      PREP          в   
66806                                                NaN      PRED      плохо   
66807                                                NaN      PRED      плохо   
66828                                                NaN      CONJ          а   

      my_pos  
56832   PRCL  
56835   PRCL  
56836   NOUN  
56837   PRCL  
56841   PRCL  
...      ...  
66651   PRED  
66671   PREP  
66806   PRED  
66807   PRED  
66828   CONJ  

[288 rows x 8 columns]

In [167]:
others = others.groupby(['lemma', 'pos'], as_index=False).agg({'meaning_ru': ', '.join})

In [169]:
others['pos'] = '<' + others['pos'].astype(str) + '>'

In [170]:
others['meaning_ru'] = '# ' + others['meaning_ru'].astype(str)

In [171]:
others

lemma     pos      meaning_ru
0         -к1а  <CONJ>           # как
1       -к1ада  <CONJ>           # как
2         -кьа  <PRCL>         # пусть
3       -хоохъ  <PRCL>          # ведь
4       а2гьаа  <intj>           # ага
..         ...     ...             ...
272     эдидаа  <CONJ>           # что
273    эдидала  <PRCL>           # все
274  экъвадала  <PRED>           # нет
275       эла2  <PREP>             # в
276   эщлъихъа  <PRED>  # плохо, плохо

[277 rows x 3 columns]

In [172]:
others.to_csv('others.txt', sep=' ', index=False, header=False)